### ISSUE TITLE
[Feature] Add versioned historical training dataset records

### ISSUE THREAD

**Issue by hnnng-w** [hnnng-w/skatmind#76]
## Context

`skat-ai` can now:

* represent a complete normal-play historical game;
* reconstruct exactly 30 information-safe decision snapshots;
* review all eligible historical decisions through the existing immediate
  recommendation workflow.

The repository does not yet provide a versioned training or evaluation dataset
record.

A historical review output is not automatically training data. A reusable
dataset representation also requires:

* stable dataset and record identities;
* source provenance;
* an explicit target definition;
* a versioned feature-generation contract;
* an explicit train, validation, or test partition;
* deterministic decision-sample generation;
* duplicate and partition-leakage protection.

This issue adds the first bounded training-data representation for complete
normal-play historical games.

It does not train a machine-learning model.

## Goals

* add a versioned public training-dataset input workflow;
* wrap complete historical games with stable provenance and partition metadata;
* deterministically generate one decision sample for each historical play;
* separate stable traceability metadata from model-facing decision features;
* use the actual historical card as the explicit supervised target;
* prevent final outcomes, future plays, and stable player identities from
  leaking into model-facing features;
* reject duplicate games and cross-partition duplication;
* establish the foundation for later statistics, opponent modeling, and
  machine-learning decisions.

## Public input workflow

Add an alternative top-level input branch:

```json
{
  "training_dataset_input": {
    "schema_version": 1,
    "dataset_id": "online-games-2026",
    "dataset_version": "1",
    "feature_generation_version": 1,
    "target": "actual_card_played",
    "records": []
  }
}
```

A file containing `training_dataset_input` is a training-dataset workflow.

It must not be combined with:

* `historical_game_input`;
* position-analysis fields;
* list-performance input branches;
* `impossible_null_settlement`;
* player-profile or opponent-policy settings.

Existing position and historical-game inputs must remain backward compatible.

## Dataset identity and versioning

Require:

* `schema_version` equal to `1`;
* a non-empty `dataset_id`;
* a non-empty `dataset_version`;
* `feature_generation_version` equal to `1`;
* `target` equal to `actual_card_played`;
* at least one dataset record.

String identifiers:

* are opaque and case-sensitive;
* must not be blank;
* must not contain leading or trailing whitespace;
* must not be generated, trimmed, or normalized by the engine.

The versions identify the public dataset structure and feature-generation
contract. They are not package versions.

## Dataset records

Each record uses:

```json
{
  "record_id": "record-001",
  "partition": "train",
  "provenance": {
    "source_type": "online_platform",
    "source_name": "Example platform",
    "source_record_id": "platform-game-123"
  },
  "historical_game": {}
}
```

Requirements:

* every `record_id` is unique within the dataset;
* every historical `game_id` is unique within the dataset;
* each record contains one complete historical game using the existing
  version-1 historical-game structure;
* historical games are validated and replayed through the existing
  historical-game implementation;
* the dataset workflow must not create a second historical-game validator.

Supported `partition` values:

* `train`;
* `validation`;
* `test`.

The engine must preserve record input order.

## Provenance

Every record requires a `provenance` object.

Require:

* `source_type`;
* `source_name`.

Supported `source_type` values:

* `online_platform`;
* `manual_entry`;
* `imported_file`;
* `synthetic`;
* `other`.

Optional provenance fields:

* `source_record_id`;
* `collected_at`;
* `notes`.

Requirements:

* all supplied strings are non-empty and have no leading or trailing
  whitespace;
* `collected_at`, when supplied, is a valid RFC 3339 date-time;
* unknown provenance properties are rejected;
* provenance is preserved unchanged in canonical output;
* provenance is traceability metadata and must not become a model-facing
  feature.

When `source_record_id` is supplied, the tuple of:

* `source_type`;
* `source_name`;
* `source_record_id`

must be unique within the dataset.

## Duplicate and partition-leakage protection

Reject:

* duplicate `record_id` values;
* duplicate historical `game_id` values;
* duplicate complete provenance identities;
* the same historical game in more than one partition;
* the same source record in more than one partition.

A duplicate is rejected even when the duplicate records use the same
partition.

This issue does not require player-disjoint partitions. Whether the same player
may occur across train, validation, and test partitions remains a later product
decision.

## Deterministic sample generation

Every valid normal-play historical game produces exactly 30 decision samples.

Generate samples from the existing information-safe historical decision
snapshots.

Do not:

* independently replay the game again;
* read the complete future play sequence while constructing an earlier sample;
* require historical recommendation or review generation;
* run simulation;
* run the recommender;
* assign decision-quality labels.

Records therefore produce:

```text
sample_count = record_count * 30
```

Sample order is:

1. dataset record input order;
2. historical `decision_index` from `1` through `30`.

Repeated conversion of the same dataset input must produce byte-stable
structured output.

## Sample structure

Each generated sample contains three separate areas:

```json
{
  "sample_id": "record-001:1",
  "metadata": {},
  "features": {},
  "label": {
    "target": "actual_card_played",
    "card": "CJ"
  }
}
```

### Sample identity

`sample_id` is deterministically derived as:

```text
record_id + ":" + decision_index
```

Requirements:

* sample IDs are unique within the dataset;
* no random identifier is generated;
* decision indices remain `1..30`;
* sample identities remain stable across repeated conversion.

### Metadata

Metadata contains the traceability information that must not be treated as a
model feature:

* `dataset_id`;
* `dataset_version`;
* `record_id`;
* `source_game_id`;
* `partition`;
* `decision_index`;
* `trick_number`;
* `play_index`;
* `acting_player_id`;
* `acting_seat`;
* `acting_side`;
* provenance reference or preserved provenance.

Stable player and source identities belong only to metadata.

### Features

Features represent only the state available immediately before the historical
play.

Use local relative positions instead of stable player identities:

* `me`;
* `left`;
* `right`.

The feature object must include the relevant decision-time information from the
existing snapshot, including:

* contract and canonical public declaration;
* acting position and side;
* own remaining hand;
* legal cards;
* current trick using relative player positions;
* completed tricks using relative player positions;
* prior completed-trick point state;
* opponent hand sizes for `left` and `right`;
* skat visibility and safely known skat cards;
* decision-time visible matadors;
* public exposed cards using relative player positions.

The exact typed structure may follow repository conventions, but:

* no stable player ID may appear inside `features`;
* no dataset, record, source, or platform identifier may appear inside
  `features`;
* no future play may appear;
* no final points may appear;
* no final winner may appear;
* no achieved future Schneider or Schwarz result may appear;
* no game-value result may appear;
* no overbid outcome may appear;
* no settlement may appear;
* no recommendation may appear;
* no decision-quality value may appear.

The actual played card remains one legal card in `own_hand` and `legal_cards`,
because those were available before the decision. Its selection as the target
must appear only in the separate `label` object.

### Label

The only supported target in version 1 is:

```json
{
  "target": "actual_card_played",
  "card": "CJ"
}
```

Requirements:

* the label card equals the historical actual card;
* it belongs to the pre-play own hand;
* it belongs to the legal-card set;
* it must not already appear in the pre-play current trick;
* no recommendation card is used as a target;
* no decision-quality category is used as a target;
* no final game result is used as a target.

Reject every other target value.

## Training-dataset output

Add a dedicated stable output branch:

```json
{
  "input_file": "examples/training_dataset_normal_play.json",
  "training_dataset_summary": {
    "schema_version": 1,
    "dataset_id": "online-games-2026",
    "dataset_version": "1",
    "feature_generation_version": 1,
    "target": "actual_card_played",
    "record_count": 1,
    "sample_count": 30,
    "partition_counts": {},
    "records": []
  }
}
```

`partition_counts` must always expose:

* `train`;
* `validation`;
* `test`.

Each partition exposes:

* `record_count`;
* `sample_count`.

Counts must reconcile with every record and sample.

Each output record contains:

* `record_id`;
* `partition`;
* preserved provenance;
* `source_game_id`;
* the canonical validated historical-game record;
* `sample_count`;
* exactly 30 decision samples.

Do not include a historical game review or recommendation output.

## CLI behavior

The existing command must accept a training-dataset file:

```powershell
python main.py --input examples/training_dataset_normal_play.json
```

Requirements:

* normal output prints dataset ID and version;
* normal output prints record and sample totals;
* normal output prints counts for train, validation, and test partitions;
* `--quiet --output` writes the structured result without successful stdout;
* historical snapshot and review flags are rejected;
* position-analysis, simulation, recommendation, comparison, policy, profile,
  sample-count, and random-seed options are rejected rather than ignored;
* existing position and historical workflows remain unchanged.

## Schemas

Create focused schemas, preferably:

```text
schemas/training_dataset.schema.json
schemas/training_dataset_output.schema.json
```

Integrate them with the public input and output schemas.

JSON Schema should enforce structural rules.

Runtime validation remains authoritative for:

* duplicate identities;
* partition leakage;
* complete historical-game validity;
* deterministic snapshot conversion;
* sample-count reconciliation;
* relative feature conversion;
* feature information safety;
* label legality;
* cross-record output reconciliation.

## Tests

Add or update focused tests covering at least the following.

### Valid datasets

* one train record;
* validation and test records;
* multiple records across different partitions;
* preserved record order;
* exactly 30 samples per record;
* deterministic repeated conversion;
* canonical structured output;
* all three partition-count entries.

### Identity and provenance

Reject:

* unsupported schema or feature-generation versions;
* blank or padded dataset IDs;
* blank or padded dataset versions;
* blank or padded record IDs;
* duplicate record IDs;
* duplicate historical game IDs;
* missing provenance;
* invalid provenance source types;
* blank provenance values;
* malformed `collected_at`;
* unknown provenance fields;
* duplicate source-record identities.

### Partitions

Reject:

* unknown partition values;
* the same game in multiple partitions;
* the same source record in multiple partitions.

Verify that:

* record counts reconcile by partition;
* sample counts reconcile by partition;
* total counts reconcile with all records.

### Features and labels

Verify for all 30 samples:

* stable sample IDs;
* chronological indices;
* stable identities are present only in metadata;
* features use only `me`, `left`, and `right`;
* own hands and legal cards match the source snapshots;
* current and completed tricks are temporally correct;
* hidden cards and future plays are absent;
* final results and settlement are absent;
* recommendations and quality values are absent;
* the actual-card label is present in own hand and legal cards;
* the target is exactly `actual_card_played`.

### Contract and visibility coverage

Cover:

* Suit;
* Grand;
* Null;
* declarer and defender decisions;
* Hand-game skat redaction;
* non-Hand declarer skat knowledge;
* ouvert public-card visibility.

Ouvert records are valid training-data records because snapshot generation does
not require the currently unsupported recommendation simulation.

### CLI and compatibility

* training-dataset input is detected correctly;
* human-readable output shows totals and partitions;
* quiet JSON output is schema-valid;
* unsupported overrides are rejected;
* existing historical-game workflows remain unchanged;
* existing position workflows remain unchanged.

## Example and generated-output validation

Add:

```text
examples/training_dataset_normal_play.json
```

The example may wrap the existing historical Grand fixture as one train record.

Add one deterministic generated-output scenario for the new workflow.

Increase the documented generated-output count from `26` to `27`.

## Documentation

Create:

```text
docs/training_data.md
```

Update relevant links and sections in:

* `README.md`;
* `docs/architecture.md`;
* `docs/input_json.md`;
* `docs/output_json.md`;
* `docs/historical_games.md`;
* `docs/historical_decision_snapshots.md`;
* `docs/historical_game_review.md`;
* `docs/examples.md`;
* `docs/schema_validation.md`;
* `docs/requirements_traceability.md`;
* `docs/v1_scope.md` only where clarification is required.

Document clearly that:

* this workflow stores and derives training/evaluation records;
* it does not train a model;
* the actual historical card is the only version-1 target;
* recommendations and decision-quality values are not training labels;
* provenance and stable identities are metadata, not model-facing features;
* partitions are explicit and duplicates across partitions are rejected;
* player-disjoint partitioning is not yet defined;
* only normally completed historical games are currently supported;
* later historical end reasons require separate support.

Update training-data representation from `not_supported` to
`partially_supported`.

Do not mark it fully supported while historical end reasons and any approved
v1 training-data gaps remain incomplete.

## Non-goals

This issue must not implement:

* machine-learning model training;
* model selection or hyperparameter tuning;
* learned opponent models;
* player-statistic aggregation;
* automatic player-disjoint partitioning;
* dataset shuffling or balancing;
* database storage;
* remote dataset upload or download;
* review-quality labels as training targets;
* recommendation cards as training targets;
* claims or concessions;
* additional historical end reasons;
* auction events;
* list, series, or tournament aggregation;
* four-player tables.

## Acceptance criteria

* A separate versioned training-dataset input workflow exists.
* Every record contains validated provenance, partition metadata, and one
  complete historical game.
* Dataset, record, game, and source identities are validated.
* Duplicate games and source records are rejected.
* Cross-partition duplication is rejected.
* Every valid record deterministically produces exactly 30 decision samples.
* Samples preserve stable traceability metadata separately from features.
* Model-facing features contain no stable player or source identities.
* Features contain only decision-time information.
* The only supported target is the legal historical actual card.
* Final outcomes, recommendations, quality values, and settlement do not leak
  into features or labels.
* Train, validation, and test counts reconcile exactly.
* Existing historical and position workflows remain backward compatible.
* Input and output schemas cover the new branch.
* One public example and generated-output scenario cover the workflow.
* Generated-output validation covers 27 scenarios.
* Relevant tests and documentation are updated.
* Training-data representation is documented as partially supported.
* The complete project check passes.
* The package version remains `0.6.0`.
* No tag or GitHub Release is created.

**Comment by hnnng-w** [comment]
Implemented and merged into `main`.

Added a separate versioned training and evaluation dataset workflow for complete normal-play historical games.

The implementation includes:

* stable dataset and record identities;
* required source provenance;
* explicit `train`, `validation`, and `test` partitions;
* duplicate record, historical game, and source-record protection;
* rejection of cross-partition game and source-record duplication;
* deterministic generation of exactly 30 decision samples per historical game;
* separate traceability metadata, model-facing features, and supervised labels;
* relative `me`, `left`, and `right` references inside model-facing features;
* `actual_card_played` as the only version-1 target;
* exclusion of future plays, hidden hands, final outcomes, settlement, recommendations, and decision-quality values;
* support for ouvert records without invoking simulation;
* dedicated input and output schemas;
* CLI support with strict rejection of unrelated analysis options;
* a public training-dataset example;
* a 27th deterministic generated-output scenario.

The workflow reuses the existing historical-game validator, replay, and information-safe snapshot generator. It does not train a machine-learning model.

Validation completed successfully:

* Ruff passed;
* input and example schema validation passed;
* all 27 generated-output scenarios passed;
* all 2,302 pytest tests passed;
* `git diff --check` passed.

Training-data support remains intentionally partial because only normally completed historical games are currently supported and player-disjoint partitioning has not yet been defined.

The package version remains `0.6.0`.

Closing as completed.
